File: solutions/prolog/twelve-days/1/twelve_days.pl
% Dedicated to Shree DR.MDD

ordinal(1, "first").
ordinal(2, "second").
ordinal(3, "third").
ordinal(4, "fourth").
ordinal(5, "fifth").
ordinal(6, "sixth").
ordinal(7, "seventh").
ordinal(8, "eighth").
ordinal(9, "ninth").
ordinal(10, "tenth").
ordinal(11, "eleventh").
ordinal(12, "twelfth").

present(1, "a Partridge in a Pear Tree").
present(2, "two Turtle Doves").
present(3, "three French Hens").
present(4, "four Calling Birds").
present(5, "five Gold Rings").
present(6, "six Geese-a-Laying").
present(7, "seven Swans-a-Swimming").
present(8, "eight Maids-a-Milking").
present(9, "nine Ladies Dancing").
present(10, "ten Lords-a-Leaping").
present(11, "eleven Pipers Piping").
present(12, "twelve Drummers Drumming").

connector(1, "") :- !.
connector(2, ", and ") :- !.
connector(_, ", ").

build_phrase(DayNum, Prior, Combined) :-
    present(DayNum, Txt),
    connector(DayNum, Join),
    format(string(Combined), "~s~s~s", [Txt, Join, Prior]).

song_line(Count, FullLine) :-
    numlist(1, Count, Indices),
    foldl(build_phrase, Indices, "", FullLine).

verse(Day, Output) :-
    ordinal(Day, Word),
    song_line(Day, FullSong),
    format(string(Output), "On the ~s day of Christmas my true love gave to me: ~s.", [Word, FullSong]).

recite(Only, Result) :- recite(Only, Only, Result).
recite(Start, End, AllLines) :-
    numlist(Start, End, Days),
    maplist(verse, Days, AllLines).

File: solutions/java/twelve-days/1/src/main/java/TwelveDays.java
import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
class TwelveDays {
    private final String TEMPLATE = "On the %s day of Christmas my true love gave to me: %s.\n"
    , SUFFIX = "th"
    , FIRST_GIFT = "a Partridge in a Pear Tree";
    private final String[] DAYS = {"first", "second", "third", "four", "fif", "six", "seven", "eigh", "nin", "ten", "eleven", "twelf"}
    , GIFTS = {"twelve Drummers Drumming", "eleven Pipers Piping", "ten Lords-a-Leaping", "nine Ladies Dancing", "eight Maids-a-Milking", "seven Swans-a-Swimming", "six Geese-a-Laying", "five Gold Rings", "four Calling Birds", "three French Hens", "two Turtle Doves"};
    
    String verse(int verseNumber) {
        return String.format(TEMPLATE
            , DAYS[verseNumber - 1] + (verseNumber < 4 ? "" : SUFFIX)
            , verseNumber == 1 ? FIRST_GIFT : Arrays.stream(GIFTS).skip(GIFTS.length - verseNumber + 1).collect(Collectors.joining(", ")) + ", and " + FIRST_GIFT);
    }
    String verses(int startVerse, int endVerse) {
        return IntStream.rangeClosed(startVerse, endVerse).mapToObj(this::verse).collect(Collectors.joining("\n"));
    }
    
    String sing() {
        return verses(1, DAYS.length);
    }
}